src/jit/stacklevelsetter.cpp

```diff
@@ -0,0 +1,266 @@
+    }
+#if !FEATURE_FIXED_OUT_ARGS
+
+    if (framePointerRequired && !comp->codeGen->isFramePointerRequired())
+    {
+        JITDUMP("framePointerRequired is not set when it is required\n");
+        comp->codeGen->resetWritePhaseForFramePointerRequired();
+        comp->codeGen->setFramePointerRequired(true);
+    }
+#endif // !FEATURE_FIXED_OUT_ARGS
+    assert(maxStackLevel <= comp->fgGetPtrArgCntMax());
+    if (maxStackLevel != comp->fgGetPtrArgCntMax())
+    {
+        JITDUMP("fgPtrArgCntMax was calculated wrong during the morph, the old value: %u, the right value: %u.\n",
+                comp->fgGetPtrArgCntMax(), maxStackLevel);
+        comp->fgSetPtrArgCntMax(maxStackLevel);
+    }
+}
+
+//------------------------------------------------------------------------
+// ProcessBlock: Do stack level calculations for one block.
+//
+// Notes:
+//   Block starts and ends with an empty outgoing stack.
+//   Nodes in blocks are iterated in the reverse order to memorize put_arg_stk
+//   stack sizes.
+//
+// Arguments:
+//   block - the block to process.
+//
+void StackLevelSetter::ProcessBlock(BasicBlock* block)
+{
+    assert(currentStackLevel == 0);
+    LIR::ReadOnlyRange& range = LIR::AsRange(block);
+    for (auto i = range.rbegin(); i != range.rend(); ++i)
+    {
+        GenTree* node = *i;
+        if (NodePutsOnStack(node))
+        {
+            GenTreePutArgStk* putArg   = static_cast<GenTreePutArgStk*>(node);
+            unsigned          numSlots = putArgNumSlots[putArg];
+            SubStackLevel(numSlots);
+        }
+
+#if !FEATURE_FIXED_OUT_ARGS
+        // Set throw blocks incoming stack depth for x86.
+        bool useThrowHlpBlk = !comp->opts.compDbgCode;
+        bool operMightThrow = ((node->gtFlags & GTF_EXCEPT) != 0);
+        if (!framePointerRequired && useThrowHlpBlk && operMightThrow)
+        {
+            SetThrowHelperBlocks(node, block);
+        }
+#endif // !FEATURE_FIXED_OUT_ARGS
+
+        if (node->IsCall())
+        {
+            GenTreeCall* call = node->AsCall();
+
+            unsigned usedStackSlotsCount = PopArgumentsFromCall(call);
+#if defined(UNIX_X86_ABI)
+            assert(call->fgArgInfo->GetStkSizeBytes() == usedStackSlotsCount * TARGET_POINTER_SIZE);
+            call->fgArgInfo->SetStkSizeBytes(usedStackSlotsCount * TARGET_POINTER_SIZE);
+#endif // UNIX_X86_ABI
+        }
+    }
+    assert(currentStackLevel == 0);
+}
+
+#if !FEATURE_FIXED_OUT_ARGS
+//------------------------------------------------------------------------
+// SetThrowHelperBlocks: Set throw helper blocks incoming stack levels targeted
+//                       from the node.
+//
+// Notes:
+//   one node can target several helper blocks.
+//
+// Arguments:
+//   node - the node to process;
+//   block - the source block for the node.
+void StackLevelSetter::SetThrowHelperBlocks(GenTree* node, BasicBlock* block)
+{
+    // Check that it uses throw block, find its kind, find the block, set level.
```

Comment:
How do we ensure this set is complete? E.g., what about GT_DIV generating SCK_DIV_BY_ZERO?

Reply:
> How do we ensure this set is complete? 

Now we do not check that. I think we can have a counter that says how many incoming block each throw helper block has, increase it when do `fgAddCodeRef` and decrease it in stack level setter. Then do assert that it is equal to 0 after the phase.

> E.g., what about GT_DIV generating SCK_DIV_BY_ZERO?

SCK_DIV_BY_ZERO is not used on x86/x64.

Reply:
Fixed.